File: 1013-fibonacci-number/1013-fibonacci-number.go
// using tail recursion and accumolators

// time complexity O(n) and space complexity O(n)
func fib(n int) int {
	return findFib(n, 0, 1)
}

func findFib(n int, a int, b int) int {
	if n == 0 {
		return a
	}

	if n == 1 {
		return b
	}

	return findFib(n-1, b, a+b)
}

File: 1013-fibonacci-number/1013-fibonacci-number.cpp
#include "iostream"
using namespace std;

// Time complexity O(n) and space O(n)
class Solution {
public:
    int fib(int n) {
        if (n == 0 || n == 1)
            return n;
        int last = fib(n - 1);
        int secondLast = fib(n - 2);

        return last + secondLast;
    }
};